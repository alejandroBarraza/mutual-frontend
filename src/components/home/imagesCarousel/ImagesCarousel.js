import React from 'react';
import './imagesCarousel.css';
import { Carousel } from 'react-carousel-minimal';
import { useQuery, gql } from '@apollo/client';
import CircularProgress from '@mui/material/CircularProgress';

/**
 * Graphql query to the database.
 */
export const IMAGES = gql`
    query GetCarrusel {
        carrusel {
            titulo
            descripcion
            imagenes {
                url
            }
        }
    }
`;

/**
 * This component returns a carousel of images from the main page.
 * @returns Carousel of images.
 */
export const ImagesCarousel = () => {
    /**
     * Database query.
     * @param loading indicates if data are still being downloaded.
     * @param error indicates if any error occurred.
     * @param data the data returned by the database.
     */
    const { loading, error, data } = useQuery(IMAGES);

    // Images of the Carrousel
    const images = data
        ? data.carrusel === null
            ? [
                  {
                      image: 'https://kinsta.com/es/wp-content/uploads/sites/8/2017/08/error-de-conexi%C3%B3n-base-de-datos.png',
                  },
              ]
            : data.carrusel.imagenes.map((img) => {
                  return { image: `${img.url}` };
              })
        : [];

    // Carousel Title
    const titleCarousel = data
        ? data.carrusel === null
            ? 'No existen datos disponibles'
            : data.carrusel.titulo
        : 'Loading...';

    // Carousel Description
    const descriptionCarousel = data
        ? data.carrusel === null
            ? 'No existen datos disponibles'
            : data.carrusel.descripcion
        : 'Loading...';

    // If there is an error, an error message is displayed.
    if (error) {
        return <p> Error en la base de datos. </p>;
    }

    return (
        <div className='container__gallery'>
            <div className='gallery'>
                <div className='gallery__text'>
                    <h1 id='titleCarousel'>{titleCarousel}</h1>
                    <p id='descriptionCarousel' className='text'>
                        {descriptionCarousel}
                    </p>
                </div>
                <div id='gallery__images' className='gallery__images'>
                    {
                        // If it is loading it returns a Progress Component, otherwise it returns the Component with the corresponding data.
                        loading ? (
                            <div id='loading__icon' className='loading__icon'>
                                <CircularProgress />
                            </div>
                        ) : (
                            <Carousel
                                data={images}
                                time={5000}
                                radius='16px'
                                slideNumber={false}
                                slideNumberStyle={{
                                    fontSize: '20px',
                                    fontWeight: 'bold',
                                }}
                                captionPosition='bottom'
                                automatic={true}
                                dots={true}
                                pauseIconColor='white'
                                pauseIconSize='40px'
                                slideBackgroundColor='darkgrey'
                                slideImageFit='cover'
                                thumbnails={false}
                                thumbnailWidth='100px'
                                style={{
                                    textAlign: 'center',
                                    width: '100%',
                                    maxWidth: '600px',
                                    maxHeight: '600px',
                                    margin: 'auto auto',
                                    overflowY: 'true',
                                }}
                            />
                        )
                    }
                </div>
            </div>
        </div>
    );
};
