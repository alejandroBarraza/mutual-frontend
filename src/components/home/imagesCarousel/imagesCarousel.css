.container__gallery {
  width: 100%;
  background-color: #e5e5e5;
  margin: 0;
}

.gallery {
  max-width: 1200px;
  width: 95%;
  display: flex;
  flex-direction: row-reverse;
  justify-content: center;
  margin: 0 auto;
  align-items: center;
  padding: 5% 0;
  gap: 1rem;
  background-color: #e5e5e5;
}

.gallery__text {
  width: 40%;
  height: 95%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-content: center;
}

.gallery__images {
  display: flex;
  width: 60%;
  height: 100%;
  padding: auto auto;
}

.text {
  font-size: 1.3rem;
  text-align: left;
}

.loading__icon {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 100%;
}

@media (max-width: 1000px) {
  .gallery {
    flex-direction: column;
    height: 100%;
    width: 100%;
    margin: 0;
  }
  .gallery__text {
    width: 95%;
    height: auto;
  }
  .gallery__images {
    width: 95%;
    height: auto;
    padding-bottom: 20px;
  }
  .text {
    font-size: 1rem;
    text-align: center;
  }
  .titleCarousel {
    text-align: center;
  }
}
