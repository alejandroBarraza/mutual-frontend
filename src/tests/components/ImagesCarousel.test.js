import { render, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { IMAGES, ImagesCarousel } from '../../components/home/imagesCarousel/ImagesCarousel';
import { CircularProgress } from '@material-ui/core';

describe('test in <ImagesCarousel /> component', () => {
   const loading = 'Loading...';
   const mocks = [
      {
         request: {
            query: IMAGES,
            variables: {
            },
         },
         result: {
            data: {
               carrusel: {
                  titulo: "Titulo del Carrusel",
                  descripcion: "Suspendisse potenti. Praesent sed tristique turpis. Aliquam quis ullamcorper nulla, eu dignissim urna. Fusce id quam eget sem pulvinar euismod et vel velit. Cras a sapien sit amet nulla lobortis ultricies. Pellentesque quis risus nec orci tincidunt fringilla. Nulla ut consequat libero, in gravida augue.",
                  imagenes: [
                     {
                        url: "https://res.cloudinary.com/dkfgeruxd/image/upload/v1633210266/IMG_20210917_210700_e0ed0f7c2f.jpg"
                     },
                  ]
               }
            },
         },
      },
   ];

   test('renders loading correctly', () => {
      const { container } = render(
         <MockedProvider mocks={mocks}>
            <ImagesCarousel />
         </MockedProvider>
      );
      const title = JSON.stringify(container.getElementsByTagName('h1').namedItem('titleCarousel').innerHTML);
      const description = JSON.stringify(container.getElementsByTagName('p').namedItem('descriptionCarousel').innerHTML);
      const circle = container.getElementsByTagName('div').namedItem('loading__icon').innerHTML;
      const circularProgress = render(< CircularProgress />).container.innerHTML;

      expect(container).toMatchSnapshot();
      expect(title).toMatch(loading);
      expect(description).toMatch(loading);
      expect(circle).toMatch(circularProgress);
   });

   test('renders loaded correctly', async () => {
      const { container } = render(
         <MockedProvider mocks={mocks}>
            <ImagesCarousel />
         </MockedProvider>
      );
      await waitFor(() => new Promise((res) => setTimeout(res, 0)));

      const title = JSON.stringify(container.getElementsByTagName('h1').namedItem('titleCarousel').innerHTML);
      const description = JSON.stringify(container.getElementsByTagName('p').namedItem('descriptionCarousel').innerHTML);
      const images = container.getElementsByTagName('div').namedItem('gallery__images').innerHTML;
      const circularProgress = render(< CircularProgress />).container.innerHTML;

      expect(container).toMatchSnapshot();
      expect(title).toMatch(mocks[0].result.data.carrusel.titulo);
      expect(description).toMatch(mocks[0].result.data.carrusel.descripcion);
      expect(images).not.toMatch(circularProgress);
   });

   // test('renders loaded with error', async () => {

   //    const error = 'No existen datos disponibles';
   //    const mocks_2 = [
   //       {
   //          request: {
   //             query: IMAGES,
   //          },
   //          error: new Error('No existen datos disponibles')
   //       },
   //    ];

   //    const { container } = render(
   //       <MockedProvider mocks={mocks_2}>
   //          <ImagesCarousel />
   //       </MockedProvider>
   //    );
   //    await waitFor(() => new Promise((res) => setTimeout(res, 0)));

   //    const title = JSON.stringify(container.getElementsByTagName('h1').namedItem('titleCarousel').innerHTML);
   //    const description = JSON.stringify(container.getElementsByTagName('p').namedItem('descriptionCarousel').innerHTML);
   //    const images = container.getElementsByTagName('div').namedItem('gallery__images').innerHTML;
   //    const circularProgress = render(< CircularProgress />).container.innerHTML;

   //    expect(container).toMatchSnapshot();
   //    expect(title).toMatch(error);
   //    expect(images).not.toMatch(circularProgress);
   // });

})


