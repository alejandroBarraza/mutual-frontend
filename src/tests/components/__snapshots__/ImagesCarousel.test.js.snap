// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <ImagesCarousel /> component renders loaded correctly 1`] = `
<div>
  <div
    class="container__gallery"
  >
    <div
      class="gallery"
    >
      <div
        class="gallery__text"
      >
        <h1
          id="titleCarousel"
        >
          Titulo del Carrusel
        </h1>
        <p
          class="text"
          id="descriptionCarousel"
        >
          Suspendisse potenti. Praesent sed tristique turpis. Aliquam quis ullamcorper nulla, eu dignissim urna. Fusce id quam eget sem pulvinar euismod et vel velit. Cras a sapien sit amet nulla lobortis ultricies. Pellentesque quis risus nec orci tincidunt fringilla. Nulla ut consequat libero, in gravida augue.
        </p>
      </div>
      <div
        class="gallery__images"
        id="gallery__images"
      >
        <div
          class="box"
          style="text-align: center; width: 100%; max-width: 600px; max-height: 600px; margin: auto auto; overflow-y: true;"
        >
          <div
            style="max-width: 600px; max-height: 400px;"
          >
            <div>
              <div
                class="carousel-container"
                style="max-width: 600px; height: 400px; background-color: darkgrey; border-radius: 16px;"
              >
                <div
                  class="carousel-item fade"
                  style="max-width: 600px; max-height: 400px; display: block;"
                >
                  <img
                    class="carousel-image"
                    src="https://res.cloudinary.com/dkfgeruxd/image/upload/v1633210266/IMG_20210917_210700_e0ed0f7c2f.jpg"
                    style="border-radius: 16px; object-fit: cover;"
                  />
                  <div
                    class="carousel-caption-bottom"
                  />
                </div>
                <a
                  class="prev"
                >
                  ❮
                </a>
                <a
                  class="next"
                >
                  ❯
                </a>
                <div
                  class="dots"
                >
                  <span
                    class="dot active"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test in <ImagesCarousel /> component renders loaded with error 1`] = `
<div>
  <div
    class="container__gallery"
  >
    <div
      class="gallery"
    >
      <div
        class="gallery__text"
      >
        <h1
          id="titleCarousel"
        >
          Loading...
        </h1>
        <p
          class="text"
          id="descriptionCarousel"
        >
          Loading...
        </p>
      </div>
      <div
        class="gallery__images"
        id="gallery__images"
      >
        <div
          class="box"
          style="text-align: center; width: 100%; max-width: 600px; max-height: 600px; margin: auto auto; overflow-y: true;"
        >
          <div
            style="max-width: 600px; max-height: 400px;"
          >
            <div>
              <div
                class="carousel-container"
                style="max-width: 600px; height: 400px; background-color: darkgrey; border-radius: 16px;"
              >
                <a
                  class="prev"
                >
                  ❮
                </a>
                <a
                  class="next"
                >
                  ❯
                </a>
                <div
                  class="dots"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test in <ImagesCarousel /> component renders loading correctly 1`] = `
<div>
  <div
    class="container__gallery"
  >
    <div
      class="gallery"
    >
      <div
        class="gallery__text"
      >
        <h1
          id="titleCarousel"
        >
          Loading...
        </h1>
        <p
          class="text"
          id="descriptionCarousel"
        >
          Loading...
        </p>
      </div>
      <div
        class="gallery__images"
        id="gallery__images"
      >
        <div
          class="loading__icon"
          id="loading__icon"
        >
          <div
            class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
